src/Haskell/Ide/Engine/Plugin/Pragmas.hs

```diff
@@ -0,0 +1,126 @@
+{-# LANGUAGE OverloadedStrings #-}
+module Haskell.Ide.Engine.Plugin.Pragmas where
+
+import           Control.Lens
+import           Data.Aeson
+import qualified Data.HashMap.Strict             as H
+import           Data.Monoid                     ((<>))
+import qualified Data.Text                       as T
+import qualified GHC.Generics                    as Generics
+import           Haskell.Ide.Engine.MonadTypes   hiding (_range)
+import qualified Language.Haskell.LSP.Types      as J
+import qualified Language.Haskell.LSP.Types.Lens as J
+
+-- ---------------------------------------------------------------------
+
+pragmasDescriptor :: PluginId -> PluginDescriptor
+pragmasDescriptor plId = PluginDescriptor
+  { pluginId = plId
+  , pluginName = "Add Missing Pragmas"
+  , pluginDesc = "Provide code actions to add missing pragmas when GHC suggests this"
+  , pluginCommands =
+      [ PluginCommand "addPragma" "add the given pragma" addPragmaCmd
+      ]
+  , pluginCodeActionProvider = Just codeActionProvider
+  , pluginDiagnosticProvider = Nothing
+  , pluginHoverProvider = Nothing
+  , pluginSymbolProvider = Nothing
+  }
+
+-- ---------------------------------------------------------------------
+
+data AddPragmaParams = AddPragmaParams
+  { file   :: Uri
+  , pragma :: T.Text
+  }
+  deriving (Show, Eq, Generics.Generic, ToJSON, FromJSON)
+
+addPragmaCmd :: CommandFunc AddPragmaParams J.WorkspaceEdit
+addPragmaCmd = CmdSync $ \_vf (AddPragmaParams uri pragmaName) -> do
+  let
+    pos = (J.Position 0 0)
+    textEdits = J.List
+      [J.TextEdit (J.Range pos pos)
+                  ("{-# LANGUAGE " <> pragmaName <> " #-}\n")
+      ]
+    res = J.WorkspaceEdit
+      (Just $ H.singleton uri textEdits)
+      Nothing
+  return $ IdeResultOk res
+
+-- ---------------------------------------------------------------------
+
+codeActionProvider :: CodeActionProvider
+codeActionProvider plId docId _ _ _ (J.CodeActionContext (J.List diags) _monly) = do
+  cmds <- mapM mkCommand pragmas
+  return $ IdeResultOk cmds
+  where
+    ghcDiags = filter (\d -> d ^. J.source == Just "ghcmod") diags
+    pragmas = concatMap (\d -> findPragma (d ^. J.message)) ghcDiags
+    mkCommand pragmaName = do
+      let
+        codeAction :: J.Command -> J.CodeAction
+        codeAction cmd = J.CodeAction title (Just J.CodeActionQuickFix) (Just (J.List [])) Nothing (Just cmd)
+        title = "Add \"" <> pragmaName <> "\""
+        cmdParams = [toJSON (AddPragmaParams (docId ^. J.uri) pragmaName )]
+      cmd <- mkLspCommand plId "addPragma" title  (Just cmdParams)
+      return $ codeAction cmd
+
+-- ---------------------------------------------------------------------
+
+findPragma :: T.Text -> [T.Text]
+findPragma str = concatMap check possiblePragmas
+  where
+    check p =
+      if T.isInfixOf p str
+        then [p]
+        else []
+
+-- ---------------------------------------------------------------------
+
+possiblePragmas :: [T.Text]
+possiblePragmas =
```

Comment:
Is this the list of the only extensions GHC suggests? If not, I think we can get the list of extensions from GHC itself.

Reply:
It is the ones that I could find being suggested. 

But I think searching for all of them would be safe, so long as it does not nail performance too much, as it gets checked on every cursor movement.

Part of me says a properly implemented compiled regexp for the match might be a good idea.  Something like an alex lexer with each token being `.*PRAGMATEXT.*`.  But that is probably overkill.

Reply:
If this is called constantly for each cursor move, then I think it makes sense to keep the list cached. Either by hardcoding it here of memoizing it after calling the GHC api.

Reply:
What is the GHC API call to use? Being lazy ...

Reply:
`DynFlags.supportedLanguagesAndExtensions`

Reply:
Output from calling this:

```
["Haskell98","Haskell2010","Unsafe","Trustworthy","Safe","AllowAmbiguousTypes","NoAllowAmbiguousTypes","AlternativeLayoutRule","NoAlternativeLayoutRule","AlternativeLayoutRuleTransitional","NoAlternativeLayoutRuleTransitional","Arrows","NoArrows","AutoDeriveTypeable","NoAutoDeriveTypeable","BangPatterns","NoBangPatterns","BinaryLiterals","NoBinaryLiterals","CApiFFI","NoCApiFFI","CPP","NoCPP","ConstrainedClassMethods","NoConstrainedClassMethods","ConstraintKinds","NoConstraintKinds","DataKinds","NoDataKinds","DatatypeContexts","NoDatatypeContexts","DefaultSignatures","NoDefaultSignatures","DeriveAnyClass","NoDeriveAnyClass","DeriveDataTypeable","NoDeriveDataTypeable","DeriveFoldable","NoDeriveFoldable","DeriveFunctor","NoDeriveFunctor","DeriveGeneric","NoDeriveGeneric","DeriveLift","NoDeriveLift","DeriveTraversable","NoDeriveTraversable","DerivingStrategies","NoDerivingStrategies","DisambiguateRecordFields","NoDisambiguateRecordFields","DoAndIfThenElse","NoDoAndIfThenElse","DoRec","NoDoRec","DuplicateRecordFields","NoDuplicateRecordFields","EmptyCase","NoEmptyCase","EmptyDataDecls","NoEmptyDataDecls","EmptyDataDeriving","NoEmptyDataDeriving","ExistentialQuantification","NoExistentialQuantification","ExplicitForAll","NoExplicitForAll","ExplicitNamespaces","NoExplicitNamespaces","ExtendedDefaultRules","NoExtendedDefaultRules","FlexibleContexts","NoFlexibleContexts","FlexibleInstances","NoFlexibleInstances","ForeignFunctionInterface","NoForeignFunctionInterface","FunctionalDependencies","NoFunctionalDependencies","GADTSyntax","NoGADTSyntax","GADTs","NoGADTs","GHCForeignImportPrim","NoGHCForeignImportPrim","GeneralizedNewtypeDeriving","NoGeneralizedNewtypeDeriving","ImplicitParams","NoImplicitParams","ImplicitPrelude","NoImplicitPrelude","ImpredicativeTypes","NoImpredicativeTypes","IncoherentInstances","NoIncoherentInstances","TypeFamilyDependencies","NoTypeFamilyDependencies","InstanceSigs","NoInstanceSigs","ApplicativeDo","NoApplicativeDo","InterruptibleFFI","NoInterruptibleFFI","JavaScriptFFI","NoJavaScriptFFI","KindSignatures","NoKindSignatures","LambdaCase","NoLambdaCase","LiberalTypeSynonyms","NoLiberalTypeSynonyms","MagicHash","NoMagicHash","MonadComprehensions","NoMonadComprehensions","MonadFailDesugaring","NoMonadFailDesugaring","MonoLocalBinds","NoMonoLocalBinds","MonoPatBinds","NoMonoPatBinds","MonomorphismRestriction","NoMonomorphismRestriction","MultiParamTypeClasses","NoMultiParamTypeClasses","MultiWayIf","NoMultiWayIf","NPlusKPatterns","NoNPlusKPatterns","NamedFieldPuns","NoNamedFieldPuns","NamedWildCards","NoNamedWildCards","NegativeLiterals","NoNegativeLiterals","HexFloatLiterals","NoHexFloatLiterals","NondecreasingIndentation","NoNondecreasingIndentation","NullaryTypeClasses","NoNullaryTypeClasses","NumDecimals","NoNumDecimals","OverlappingInstances","NoOverlappingInstances","OverloadedLabels","NoOverloadedLabels","OverloadedLists","NoOverloadedLists","OverloadedStrings","NoOverloadedStrings","PackageImports","NoPackageImports","ParallelArrays","NoParallelArrays","ParallelListComp","NoParallelListComp","PartialTypeSignatures","NoPartialTypeSignatures","PatternGuards","NoPatternGuards","PatternSignatures","NoPatternSignatures","PatternSynonyms","NoPatternSynonyms","PolyKinds","NoPolyKinds","PolymorphicComponents","NoPolymorphicComponents","PostfixOperators","NoPostfixOperators","QuasiQuotes","NoQuasiQuotes","Rank2Types","NoRank2Types","RankNTypes","NoRankNTypes","RebindableSyntax","NoRebindableSyntax","RecordPuns","NoRecordPuns","RecordWildCards","NoRecordWildCards","RecursiveDo","NoRecursiveDo","RelaxedLayout","NoRelaxedLayout","RelaxedPolyRec","NoRelaxedPolyRec","RoleAnnotations","NoRoleAnnotations","ScopedTypeVariables","NoScopedTypeVariables","StandaloneDeriving","NoStandaloneDeriving","StaticPointers","NoStaticPointers","Strict","NoStrict","StrictData","NoStrictData","TemplateHaskell","NoTemplateHaskell","TemplateHaskellQuotes","NoTemplateHaskellQuotes","TraditionalRecordSyntax","NoTraditionalRecordSyntax","TransformListComp","NoTransformListComp","TupleSections","NoTupleSections","TypeApplications","NoTypeApplications","TypeInType","NoTypeInType","TypeFamilies","NoTypeFamilies","TypeOperators","NoTypeOperators","TypeSynonymInstances","NoTypeSynonymInstances","UnboxedTuples","NoUnboxedTuples","UnboxedSums","NoUnboxedSums","UndecidableInstances","NoUndecidableInstances","UndecidableSuperClasses","NoUndecidableSuperClasses","UnicodeSyntax","NoUnicodeSyntax","UnliftedFFITypes","NoUnliftedFFITypes","ViewPatterns","NoViewPatterns"]
```

Are only of these pragma-ble?